### PR TITLE
For custom cassette persisters no longer catch ValueError

### PR DESCRIPTION
**Pull request by amosjyng**
I had to spend some time digging before I realized that serialization/deserialization errors were being swallowed here

### CODE REVIEW

**Review by hartwork — COMMENTED** [review]
@amosjyng I see how this change could help on an _abstract_ level. Could you elaborate what _concrete_ practical problem this is fixing for you? What made you want to change this code? Thank in advance! :pray:

**Comment by amosjyng** [comment]
IIRC I was trying to save different keys to the recorded file, but `deserialize(data, serializer)` was throwing a ValueError because the keys were not as expected. vcrpy would continually treat the session recording as if it weren't there at all, and it took a little while to debug that it was due to this mismatch between how the session was serialized vs how it was deserialized. Hence, this PR as a suggestion to make such development easier in the future.

**Comment by hartwork** [comment]
Hi @amosjyng,

thanks for elaborating!

I just found that the documentation at https://vcrpy.readthedocs.io/en/latest/advanced.html#register-your-own-cassette-persister says that…

> [t]he `load_cassette` method must return a deserialized cassette or raise `ValueError` if no cassette is found.

I think that means that changing `except ValueError:` would break documented API. If that understanding is correct it would mean that we either (a) cannot change it or (b) need a major version bump to clearly communicate this change.

So we probably need to clarify that first — @jairhenrique  @kevin1024 what do you think, does this break the API?

Best, Sebastian

---
Notes to self for later:
- rebase onto latest `master` to get the CI involved
- class name `ExpectedCassetteError` not ideal?
- add a simple test using a mock custom cassette persistor

**Comment by kevin1024** [comment]
1. Yes, this would be a breaking change to a documented API and would require a major version bump
2. Yeah, this probably would be a quality-of-life improvement for those using vcrpy.

My guess is that most folks have *not* built their own persister, so it should only affect a minority of vcrpy users.  Has anyone built their own persister?  Honestly IIRC I just thought it was a cool idea for someone to store cassettes in a database or s3 or something but I haven't heard of anyone using it.

**Comment by hartwork** [comment]
@kevin1024 to be sure I read (2) right: You're pro fixing this and the major version bump?

**Comment by kevin1024** [comment]
Yeah,  but this single change alone probably isn’t worth releasing a version bump in and of itself.  Maybe we can batch it up with some other breaking changes like dropping older models python versions and/or old http library versions and do a major bump sometime in the future. What do you think?

**Comment by hartwork** [comment]
> Yeah, but this single change alone probably isn’t worth releasing a version bump in and of itself. Maybe we can batch it up with some other breaking changes like dropping older models python versions and/or old http library versions and do a major bump sometime in the future. What do you think?

@kevin1024 if we go the collect-more-breakage route, we either need a second master branch or we lose instant releasability on the master branch. The latter is a clear non-option to me and former is probably overkill and may hold this for later for an unclear amount of time. Given that the breakage is easy to adjust to, will likely not affect most users (as you mentioned), and will likely show directly where used in the test suite, I would vote for the simple approach and just treat this like a normal pull request except with bumping the version to 5.0.0, and continue as usual. If you like, we can merge #695 dropping Python 3.7 right after and call that the release, if you feel it's in the breaking bucket (— personally, I don't see it as breaking but it can be argued). What do you think?

**Comment by kevin1024** [comment]
Excellent points, all!  I agree with your plan.  The only thing this PR needs still is an update to the documentation to mention the new API (different exception)

**Comment by hartwork** [comment]
> Excellent points, all! I agree with your plan.

@kevin1024 cool!

> The only thing this PR needs still is an update to the documentation to mention the new API (different exception)

Yes!

I believe in total (integrating my earlier todo-notes from https://github.com/kevin1024/vcrpy/pull/681#issuecomment-1581749936 above) we need:
- a rebase onto latest `master` (e.g. to bring the CI in)
- probably two(?) new exception classes to cleanly distinct the two likely major/common cases for custom persisters, e.g.
  - `CassetteNotFoundError`, potentially inheriting from built-in `FileNotFoundError`
  - `CassetteDecodeError`, potentially inheriting from `ValueError`
- a simple test per new exception
- doc fixes at https://github.com/kevin1024/vcrpy/blob/f4316d2daea94a3baaad2aaac443c97548843e59/docs/advanced.rst?plain=1#L139

@amosjyng @kevin1024 what do you think?

PS: I can help out with any of these parts as needed.

**Comment by amosjyng** [comment]
Sounds great to me! There's no rush on my end, we can wait for whenever the next major version is due. I just opened this PR as a suggestion to hopefully save others in the future some time :)

**Comment by hartwork** [comment]
Hi @amosjyng I think it's likely that (driven by #695) there will be a natural next release before ~2023-07-01. Let's get started on the `ValueError` thing then so that we're done by then, I'd be happy to review the next iteration.

**Comment by hartwork** [comment]
@amosjyng any news on the implementation?

**Comment by amosjyng** [comment]
Sorry, I thought you guys were talking to each other about making those changes, not to me :P

I've split it into `CassetteNotFoundError` and `CassetteDecodeError` as requested. I've added tests for a custom persister that throws those errors, as well as a test for the default persister throwing a `CassetteNotFoundError`. I'm not sure how to get the default persister to throw the `CassetteDecodeError` as it's not clear to me how you'd trigger a `UnicodeEncodeError` while reading a file; should we perhaps be checking for `UnicodeDecodeError` as well?

File: docs/advanced.rst
Comment (hartwork) [inline docs/advanced.rst:139]: This should mention `CassetteDecodeError` also somehow or no one will know about that exception with reading the internal code.

File: tests/integration/test_register_persister.py
Comment (hartwork) [inline tests/integration/test_register_persister.py:76]: Could you make "stack" be "call stack"? I think it would make it a tiny bit more clear.

File: vcr/persisters/filesystem.py
Comment (hartwork) [inline vcr/persisters/filesystem.py:26]: I believe `except UnicodeEncodeError as err:` needs to be `except UnicodeDecodeError as err:`, instead.

(Even if loading were to decode and re-encoded certain bytes, I believe it's unlikely that that re-encoding would fail, if decoding worked the first time. So probably we don't need to catch both but only `UnicodeDecodeError`. What do you think?)

File: tests/integration/test_register_persister.py
Comment (hartwork) [inline tests/integration/test_register_persister.py:73]: How about a rename to `test_load_cassette_persister_exception_handling`? It think something like that would be a bit more clear and in line with the docstring.

File: tests/unit/test_cassettes.py
Comment (hartwork) [inline tests/unit/test_cassettes.py:42]: With `load_cassette` fixed to handle `UnicodeDecodeError` I believe this test should then be:

```python
assert Cassette.load(path=a_file) is None
```

(with no more `pytest.raises`)

File: tests/unit/test_cassettes.py
Comment (hartwork) [inline tests/unit/test_cassettes.py:40]: Just a detail: I would prefer wording "invalid encoding" (or "malformed encoding") over "invalid Unicode" here, to be more precise.

**Review by hartwork — CHANGES_REQUESTED** [review]
@amosjyng thanks for this, looking pretty good already, nice! :+1: 

I was wondering whether `UnicodeEncodeError` should actually be `UnicodeDecodeError` instead also — by now I think it's indeed mixed up. Please see my comments on details below:

File: docs/advanced.rst
Comment (amosjyng) [reply docs/advanced.rst]: Done! Feel free to change the wording further.

File: vcr/persisters/filesystem.py
Comment (amosjyng) [reply vcr/persisters/filesystem.py]: Agreed, done!

File: tests/integration/test_register_persister.py
Comment (amosjyng) [reply tests/integration/test_register_persister.py]: Done!

File: tests/integration/test_register_persister.py
Comment (amosjyng) [reply tests/integration/test_register_persister.py]: Done! I just realized I forgot to rename the function after copy-pasting, so this fixes the name collision as well.

File: tests/unit/test_cassettes.py
Comment (amosjyng) [reply tests/unit/test_cassettes.py]: A Cassette is still returned, but with zero records, same as the case where the file doesn't exist. I've modified it to be

```python
    a_cassette = Cassette.load(path=str(a_file))
    assert len(a_cassette) == 0
```

to keep in line with the asserts from `test_cassette_load` and `test_cassette_load_nonexistent`

File: tests/unit/test_cassettes.py
Comment (amosjyng) [reply tests/unit/test_cassettes.py]: Good point, done!

**Review by hartwork — APPROVED** [review]
@amosjyng thanks for the adjustments, looks great! :+1: 

I have some hope that @kevin1024 is with me on releasing 4.4.0 shortly **before** the merge (i.e. issue #724). Let's see what he thinks; approving the pull request now…

**Comment by hartwork** [comment]
@amosjyng you probably saw https://github.com/kevin1024/vcrpy/releases/tag/v5.0.0 but… your documentation changes are also live at https://vcrpy.readthedocs.io/en/latest/advanced.html#register-your-own-cassette-persister now. Thanks again! :pray:

**Comment by amosjyng** [comment]
Woohooo, that's cool to see! Thanks @hartwork ! :)

**Pull request merged**
